File: 2015/lua/4/main.lua
local md5 = require 'md5'
function main()
    if #arg < 1 then
        print("Usage: lua main.lua <input.txt>")
        os.exit(1)
    end

    local filename = arg[1]

    local file = io.open(filename, "r")

    if not file then
        print("Could not open file: " .. filename)
        os.exit(1)
    end

    local lines = {}

    for line in file:lines() do
        table.insert(lines, line)
    end

    file:close()

    local prefix = lines[1]

    local i = 0

    local leading_5_found = false

    while true do
        local hash = md5.sumhexa(prefix .. i)
        -- check if the hash starts with 5 zeros
        if hash:sub(1, 5) == "00000" and not leading_5_found then
            print("Part 1: " .. i)
            leading_5_found = true
        end

        if hash:sub(1, 6) == "000000" then
            print("Part 2: " .. i)
            break
        end

        i = i + 1
    end
end

main()

File: 2015/go/4/main.go
package main

import (
	"crypto/md5"
	"fmt"
	"os"
	"strconv"
)

func main() {
	// get command line arguments
	args := os.Args[1:]

	// check if there are any arguments
	if len(args) == 0 {
		// if no arguments, print usage
		println("Usage: go run main.go <input.txt>")
		return
	}

	filename := args[0]

	// read the file
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}

	// close the file when the function returns
	defer file.Close()

	var line string
	_, err = fmt.Fscanln(file, &line)
	if err != nil {
		panic(err)
	}

	var i = 0
	var leading_5_found = false

	for {
		var input string
		input = line + "" + strconv.FormatInt(int64(i), 10)

		// calculate the hash
		hash := md5.Sum([]byte(input))

		// check if the hash starts with 5 zeros
		if hash[0] == 0 && hash[1] == 0 && hash[2] < 16 && !leading_5_found {
			fmt.Printf("Part 1: %d\n", i)
			leading_5_found = true
		}

		// check if the hash starts with 6 zeros
		if hash[0] == 0 && hash[1] == 0 && hash[2] == 0 {
			fmt.Printf("Part 2: %d\n", i)
			break
		}

		// increment the number
		i++
	}
}
